Inferring generic array type for a structure field does not compile

### Describe the bug

I was reducing the test [vlib/v/tests/method_call_var_comp_test.v](https://github.com/vlang/v/blob/master/vlib/v/tests/method_call_var_comp_test.v) to investigate, why inferring of the generic argument didn't work in my source, and ended up with the compiler choking on a valid source code. Compiling the following file:

```go
struct EmptyStruct {
	b []int
}

fn encode_array[T](val []T) {
}

fn encode_struct[T](val T) {
	$if T is $struct {
		$for field in T.fields {
			$if field.is_array {
				encode_array(val.$(field.name))
			}
		}
	}
}

fn test_encoder() {
	encode_struct(EmptyStruct{})
}
```

ends up with the following message:

```txt
❯ v bug_encode.v
==================
/tmp/v_501/bug_encode.10979425198745433273.tmp.c:12530:6: error: call to undeclared function 'main__encode_array_T_Array_int'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
                                        main__encode_array_T_Array_int(val.b);
                                        ^
/tmp/v_501/bug_encode.10979425198745433273.tmp.c:12530:6: note: did you mean 'main__encode_array_T_int'?
/tmp/v_501/bug_encode.10979425198745433273.tmp.c:12503:22: note: 'main__encode_array_T_int' declared here
VV_LOCAL_SYMBOL void main__encode_array_T_int(Array_int val) {
                     ^
1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Expected Behavior

Compiling should succeed and the statement:

```go
encode_array(val.$(field.name))
```

should end up inferring the following function:


```go
fn encode_array[int](val []int) {
}
```

### Current Behavior

The compiler fails with the error shown above.

### Reproduction Steps

1. Save the code above to a file `bug_encode.v`.
2. Execute `v bug_encode.v`.

Or observe it at https://vosca.dev/p/79548d3a73.

### Possible Solution

This is just from me observing the error message. I might be completely wrong, because I don't know the internals of the compiler.

Remove the work `Array` from the function call, which generated for function calls to `func[T](a []T)`. Currently the called name looks like `main__encode_array_T_Array_int`. It should be `main__encode_array_T_int`, because although the input argument may be an array, the generic type infers to just an array element.

### Additional Information/Context

When I look at the generated C code, it contains the function `main__encode_array_T_int`:

```c
VV_LOCAL_SYMBOL void main__encode_array_T_int(Array_int val);

VV_LOCAL_SYMBOL void main__encode_array_T_int(Array_int val) {
}
```

However, the function call generated later calls `main__encode_array_T_Array_int`:

```c
VV_LOCAL_SYMBOL void main__encode_struct_T_main__EmptyStruct(main__EmptyStruct val) {
	#if 1
	{
		/* $for field in main.EmptyStruct.fields */ {
				FieldData field = {0};
			/* field 0 */ {
				field.name = _SLIT("b");
				field.attrs = __new_array_with_default(0, 0, sizeof(string), 0);
				field.typ = 37;
				field.unaliased_typ = 37;
				field.is_pub = false;
				field.is_mut = false;
				field.is_shared = false;
				field.is_atomic = false;
				field.is_option = false;
				field.is_array = true;
				field.is_map = false;
				field.is_chan = false;
				field.is_struct = false;
				field.is_alias = false;
				field.is_enum = false;
				field.indirections = 0;
				#if true
				{
					main__encode_array_T_Array_int(val.b);
				}
				#endif
			}
		}// $for
	}
	#endif
}
```

It appears that the `Array` comes from the argument `val`:

```go
fn encode_array[T](val []T) {
}
```

when the `encode_array` is called with the field `b`, which is `[]int`:

```go
encode_array(val.$(field.name))
```

If I remove the struct and try just a plain variable, the code compiles well and the generated C function is `main__encode_array_T_int`:

```go
fn encode_array[T](val []T) {
}

fn test_encoder() {
	i := []int{}
	encode_array(i)
}
```

Also at https://vosca.dev/p/1cd599b366.


### V version

V 0.3.4 3ac63cc

### Environment details (OS name and version, etc.)

V full version: V 0.3.4 1be798b.3ac63cc
OS: macos, macOS, 13.3.1, 22E772610a
Processor: 12 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz

getwd: /Users/prantlf/Sources/github/v-yaml
vexe: /usr/local/Cellar/vlang/0.3.4/libexec/v
vexe mtime: 2023-05-28 15:45:55

vroot: OK, value: /usr/local/Cellar/vlang/0.3.4/libexec
VMODULES: OK, value: /Users/prantlf/.vmodules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.40.1
Git vroot status: weekly.2023.21-38-g3ac63cce
.git/config present: true

CC version: Apple clang version 14.0.3 (clang-1403.0.22.14.1)
thirdparty/tcc status: thirdparty-macos-amd64 46662e20

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1115,6 +1115,12 @@ fn (mut g Gen) change_comptime_args(func ast.Fn, mut node_ ast.CallExpr, concret
 				}
 			} else if mut call_arg.expr is ast.ComptimeSelector {
 				comptime_args[i] = g.comptime_for_field_type
+				arg_sym := g.table.final_sym(call_arg.typ)
+				param_typ_sym := g.table.sym(param_typ)
+				if arg_sym.kind == .array && param_typ.has_flag(.generic)
+					&& param_typ_sym.kind == .array {
+					comptime_args[i] = g.get_generic_array_element_type(arg_sym.info as ast.Array)
+				}
 				if call_arg.expr.left.is_auto_deref_var() {
 					comptime_args[i] = comptime_args[i].deref()
 				}

--- a/vlib/v/tests/generic_comptime_test.v
+++ b/vlib/v/tests/generic_comptime_test.v
@@ -1,0 +1,26 @@
+struct EmptyStruct {
+	b []int = [1, 2, 3]
+}
+
+fn encode_array[T](val []T) []int {
+	return val
+}
+
+fn encode_struct[T](val T) []int {
+	$if T is $struct {
+		$for field in T.fields {
+			$if field.is_array {
+				return encode_array(val.$(field.name))
+			}
+		}
+	}
+	return [0]
+}
+
+fn encoder() []int {
+	return encode_struct(EmptyStruct{})
+}
+
+fn test_main() {
+	assert encoder() == [1, 2, 3]
+}

